benchmarks/nemo/tracer_advection/scripts/acc_mixed_explicit_data_movement_trans.py

```diff
@@ -0,0 +1,207 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
+
+'''A transformation script that seeks to apply OpenACC KERNELS
+directives to NEMO style code.  In order to use
+it you must first install PSyclone. See README.md in the top-level
+psyclone directory.
+
+Once you have psyclone installed, this may be used by doing:
+
+ $ psyclone -api nemo -s ./kernels_trans.py some_source_file.f90
+
+This should produce a lot of output, ending with generated
+Fortran. Note that the Fortran source files provided to PSyclone must
+have already been preprocessed (if required).
+
+The transformation script attempts to insert Kernels directives at the
+highest possible location(s) in the schedule tree (i.e. to enclose as
+much code as possible in each Kernels region). However, due to
+limitations in the PGI compiler, we must take care to exclude certain
+nodes (such as If blocks) from within Kernel regions. If a proposed
+region is found to contain such a node (by the ``valid_kernel``
+routine) then the script moves a level down the tree and then repeats
+the process of attempting to create the largest possible Kernel
+region.
+
+'''
+
+from __future__ import print_function
+from psyclone.domain.nemo.transformations import NemoAllArrayRange2LoopTrans
+from psyclone.errors import InternalError
+from psyclone.psyir.nodes import Assignment, CodeBlock, Call, Literal, Loop, \
+    ACCLoopDirective
+from psyclone.psyir.transformations import ACCUpdateTrans
+from psyclone.transformations import ACCLoopTrans, TransformationError, \
+    ACCKernelsTrans, ACCEnterDataTrans
+
+
+COLLAPSE_LOOPS = False
+EXPAND_IMPLICIT_LOOPS = False
+
+# Get the PSyclone transformations we will use
+ARRAY_RANGE_TRANS = NemoAllArrayRange2LoopTrans()
+
+
+def trans(psy):
+    '''A PSyclone-script compliant transformation function. Applies
+    OpenACC 'kernels' and 'data' directives to NEMO code.
+
+    :param psy: The PSy layer object to apply transformations to.
+    :type psy: :py:class:`psyclone.psyGen.PSy`
+    '''
+
+    print("Invokes found:\n{0}\n".format(
+        "\n".join([str(name) for name in psy.invokes.names])))
+
+    for invoke in psy.invokes.invoke_list:
+
+        sched = invoke.schedule
+        if not sched:
+            print("Invoke {0} has no Schedule! Skipping...".
+                  format(invoke.name))
+            continue
+        sched.view()
+
+        if EXPAND_IMPLICIT_LOOPS:
+            # Transform any array assignments (Fortran ':' notation)
+            # into loops.
+            for assignment in sched.walk(Assignment):
+                ARRAY_RANGE_TRANS.apply(assignment)
+
+        add_kernels(sched.children)
```

Comment:
Could/should use the add_kernels from utils.py?

Reply:
Yes.
Initially I didn't do it because the "loop independent" was added at the bottom of this call chain. But now I add the kernels and the loop directives independently in two steps and always using the utils infrastructure. The new version is slightly better because it manages to put more loop directives and collapse to 3 when possible.